### PR TITLE
MC-193: Remove hard coded table names

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | n
| BC breaks?           | n
| Tests pass?          | y
| Checkstyle issues?*  | n
| PMD issues?**        | n
| Changelog updated?   | n
| Fixed tickets        | MC-193
| DB schema updated?   | n
| Migration script?    | n
| Doc PR               |

### CODE REVIEW

File: Reader/ORM/DeltaProductReader.php [hunk 1]
```diff
@@ -13,6 +19,57 @@
...
+     * @param boolean                    $missingCompleteness
+     */
+    public function __construct(
+        ProductRepositoryInterface $repository,
+        ChannelManager $channelManager,
+        CompletenessManager $completenessManager,
+        MetricConverter $metricConverter,
+        EntityManager $entityManager,
+        TableNameBuilder $tableNameBuilder,
+        $catalogProductParamName,
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: $productClass

File: Reader/ORM/DeltaProductReader.php [hunk 2]
```diff
@@ -13,6 +19,57 @@
...
+        $categoryParamName,
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: $categoryClass

File: Reader/ORM/DeltaProductReader.php [hunk 3]
```diff
@@ -13,6 +19,57 @@
...
+        $deltaProductParamName,
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: $deltaProductClass

File: Builder/TableNameBuilder.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * Construct
+     *
+     * @param ContainerInterface $container
+     * @param ManagerRegistry    $managerRegistry
+     */
+    public function __construct(ContainerInterface $container, ManagerRegistry $managerRegistry)
+    {
+        $this->container = $container;
+        $this->managerRegistry = $managerRegistry;
```
Comment (fitn) [inline Builder/TableNameBuilder.php]: I think EntityManager is enough in fact.

File: Builder/TableNameBuilder.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $entityClassName = $this->getEntityClassName($entityParameter);
+        $classMetadata   = $this->getClassMetadata($entityClassName);
+
+        if (null !== $targetEntity) {
+            $assocMapping  = $classMetadata->getAssociationMapping($targetEntity);
+
+            if (true === $isJoin) {
+                $tableName = $assocMapping['joinTable']['name'];
+            } else {
```
Comment (fitn) [inline Builder/TableNameBuilder.php]: I think you could calculate this one and avoid a third optional parameter

File: Reader/ORM/DeltaProductReader.php [hunk 6]
```diff
@@ -49,17 +106,30 @@ protected function getIds()
...
     {
+        $productTable = $this->tableNameBuilder->getTableName($this->catalogProductParamName);
+        $completenessesTable = $this->tableNameBuilder->getTableName(
+            $this->catalogProductParamName,
+            'completenesses'
+        );
+        $categoryProductTable = $this->tableNameBuilder->getTableName(
+            $this->catalogProductParamName,
+            'categories',
+            true
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: I think you could remove this third parameter

File: Reader/ORM/DeltaProductReader.php [hunk 7]
```diff
@@ -49,17 +106,30 @@ protected function getIds()
...
+        );
+        $categoryTable = $this->tableNameBuilder->getTableName($this->categoryParamName);
+        $deltaProductExportTable = $this->tableNameBuilder->getTableName($this->deltaProductParamName);
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: $deltaProductTable

File: Reader/ORM/DeltaProductReader.php (hunk 3 above)
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: In fact, it's useless to inject these param names. We know them and they won't change.

File: Reader/ORM/DeltaProductReader.php [hunk 8]
```diff
@@ -49,17 +106,30 @@ protected function getIds()
      */
     protected function getSQLQuery($channelId, $treeId, $jobInstanceId)
     {
+        $productTable = $this->tableNameBuilder->getTableName($this->catalogProductParamName);
```
Comment (fitn) [inline Reader/ORM/DeltaProductReader.php]: You should directly call pim_catalog.entity.product.class here and same for all your param's names

File: Builder/TableNameBuilder.php (hunk 5 above)
Comment (willy-ahva) [inline Builder/TableNameBuilder.php]: I can't compute it, you could want join table name or target entity name. Can't know for you.

File: Manager/DeltaProductExportManager.php [hunk 9]
```diff
@@ -33,19 +28,25 @@ class DeltaProductExportManager
...
+     * @param ProductRepositoryInterface $productRepository Product repository
+     * @param boolean                    $productValueDelta Should we do a delta on product values
+     * @param TableNameBuilder           $tableNameBuilder  Table name builder
      */
     public function __construct(
         EntityManager $entityManager,
         ProductRepositoryInterface $productRepository,
-        $productValueDelta = false
+        $productValueDelta = false,
+        TableNameBuilder $tableNameBuilder
```
Comment (fitn) [inline Manager/DeltaProductExportManager.php]: Here you can switch parameters. No inheritance nor BC.

**Pull request merged**